File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaPage.java
package in.org.quicko.silo.client.media;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.pagination.Page;
import in.org.quicko.silo.client.pagination.PageWindow;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.Transport;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * One page of the library, navigated by the window the server echoed rather than
 * the one requested. {@link Media} builds the wire query once and every page
 * reuses it unchanged.
 */
public final class MediaPage extends Page<MediaAsset> {
  private final Transport transport;
  private final Map<String, Object> wireQuery;

  private MediaPage(
      List<MediaAsset> rows,
      int total,
      PageWindow window,
      Transport transport,
      Map<String, Object> wireQuery) {
    super(rows, total, window);
    this.transport = transport;
    this.wireQuery = wireQuery;
  }

  public List<MediaAsset> files() {
    return rows();
  }

  public Optional<MediaPage> next() {
    return next(RequestOptions.none());
  }

  public Optional<MediaPage> next(RequestOptions options) {
    return windowForNext().map(window -> load(transport, wireQuery, window, options));
  }

  public Optional<MediaPage> previous() {
    return previous(RequestOptions.none());
  }

  public Optional<MediaPage> previous(RequestOptions options) {
    return window().previous().map(window -> load(transport, wireQuery, window, options));
  }

  static MediaPage load(
      Transport transport, Map<String, Object> wireQuery, PageWindow window, RequestOptions options) {
    JsonNode body = transport.json(
        TransportRequest.get(ApiPath.media())
            .queries(wireQuery)
            .query("limit", window.limit())
            .query("offset", window.offset())
            .options(options)
            .build());

    List<MediaAsset> assets = new ArrayList<>();
    body.path("items").forEach(item ->
        assets.add(new MediaAsset(transport, MediaMapper.toRecord(item))));

    PageWindow answered = new PageWindow(
        body.path("limit").isNumber() ? body.get("limit").asInt() : window.limit(),
        body.path("offset").isNumber() ? body.get("offset").asInt() : window.offset());

    return new MediaPage(assets, body.path("total").asInt(0), answered, transport, wireQuery);
  }
}

File: packages/silo-client-dart/lib/src/media/media_page.dart
import '../pagination/page.dart';
import '../pagination/page_window.dart';
import '../request_options.dart';
import '../transport/api_path.dart';
import '../transport/page_payload.dart';
import '../transport/transport.dart';
import '../transport/transport_request.dart';
import 'media_asset.dart';
import 'media_asset_record.dart';

/// One page of `Media.list`, navigated by the window the server answered.
final class MediaPage extends Page<MediaAsset> {
  MediaPage._(this.files, int total, PageWindow window, this._transport, this._wireQuery) : super(files, total, window);

  final List<MediaAsset> files;
  final Transport _transport;
  final Map<String, Object?> _wireQuery;

  Future<MediaPage?> next([RequestOptions options = const RequestOptions()]) async {
    final window = windowForNext();
    return window == null ? null : load(_transport, _wireQuery, window, options);
  }

  Future<MediaPage?> previous([RequestOptions options = const RequestOptions()]) async {
    final window = windowForPrevious();
    return window == null ? null : load(_transport, _wireQuery, window, options);
  }

  static Future<MediaPage> load(
    Transport transport,
    Map<String, Object?> wireQuery,
    PageWindow window,
    RequestOptions options,
  ) async {
    final body = await transport.json(
      TransportRequest(
        'GET',
        ApiPath.media(),
        query: {...wireQuery, 'limit': window.limit, 'offset': window.offset},
        options: options,
      ),
    );
    final payload = PagePayload.read(body);
    return MediaPage._(
      [for (final row in payload.rows) MediaAsset(transport, MediaAssetRecord.fromJson(row))],
      payload.total,
      PageWindow(payload.limit ?? window.limit, payload.offset ?? window.offset),
      transport,
      wireQuery,
    );
  }
}
